File: RegisterForm/app/src/main/java/com/outlook/gonzasosa/registerform/MainActivity.java
package com.outlook.gonzasosa.registerform;

import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import android.widget.Toolbar;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.outlook.gonzasosa.registerform.databinding.ActivityMainBinding;
import com.outlook.gonzasosa.registerform.vm.FormViewModel;

public class MainActivity extends AppCompatActivity {
    private final String ACTIVITY_TITLE = "Register Form";
    private final String PREFERENCES_NAME = "RegisterForm";
    private final String FIRST_NAME_KEY = "FirstName";
    private final String LAST_NAME_KEY = "LastName";
    private final String AGE_KEY = "Age";
    private final String PHONE_KEY = "Phone";

    private FormViewModel viewModel;
    private ActivityMainBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = new ViewModelProvider(this).get(FormViewModel.class);
        binding = ActivityMainBinding.inflate(getLayoutInflater());

        setContentView (binding.getRoot ());

        Toolbar toolbar = findViewById(R.id.toolbar);
        if (toolbar != null)
        {
            setActionBar(toolbar);
            setTitle (ACTIVITY_TITLE);
        }

        Button btnSave = findViewById(R.id.btnSave);
        btnSave.setOnClickListener(view ->
        {
            String firstName = binding.edtName.getText().toString();
            Toast.makeText(this, firstName, Toast.LENGTH_SHORT).show();
        } );
    }

    @Override
    protected void onResume() {
        super.onResume();

        binding.edtName.setText(viewModel.FirstName);
        binding.edtLastName.setText(viewModel.LastName);
        binding.edtAge.setText(String.valueOf(viewModel.Age));
        binding.edtPhone.setText(viewModel.Phone);
    }

    @Override
    protected void onPause() {
        super.onPause();

        viewModel.FirstName = binding.edtName.getText().toString();
        viewModel.LastName = binding.edtLastName.getText().toString();
        viewModel.Age = Integer.parseInt(binding.edtAge.getText().toString());
        viewModel.Phone = binding.edtPhone.getText().toString();
    }

    @Override
    protected void onStop () {
        super.onStop ();
        savePreferences ();
    }

    private void savePreferences ()
    {
        var sharedPreferences = getSharedPreferences (PREFERENCES_NAME, MODE_PRIVATE);
        var editor = sharedPreferences.edit ();

        editor.putString (FIRST_NAME_KEY, viewModel.FirstName);
        editor.putString (LAST_NAME_KEY, viewModel.LastName);
        editor.putInt    (AGE_KEY, viewModel.Age);
        editor.putString (PHONE_KEY, viewModel.Phone);
        editor.apply ();
    }
}

File: RegisterFormKts/app/src/main/java/com/outlook/gonzasosa/registerformkts/MainActivity.kt
package com.outlook.gonzasosa.registerformkts

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.edit
import androidx.lifecycle.ViewModelProvider

import com.outlook.gonzasosa.registerformkts.databinding.ActivityMainBinding
import com.outlook.gonzasosa.registerformkts.viewmodel.FormViewModel

class MainActivity : AppCompatActivity() {
    private val registerForm: String = "RegisterForm"
    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: FormViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        viewModel = ViewModelProvider(this)[FormViewModel::class]

        setContentView (binding.root)
        setSupportActionBar(binding.toolbar)

        loadPreferences()

        binding.btnSave.setOnClickListener {
            val firstName = binding.edtName.text.toString()
            Toast.makeText(this, firstName, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onResume() {
        super.onResume()

        binding.edtName.setText (viewModel.firstName)
        binding.edtLastName.setText (viewModel.lastName)
        binding.edtAge.setText (viewModel.age.toString())
        binding.edtPhone.setText (viewModel.phone)
    }

    override fun onPause() {
        super.onPause()

        val age = if (binding.edtName.text.toString().isEmpty()) 0
        else binding.edtAge.text.toString().toInt()

        viewModel.firstName = binding.edtName.text.toString ()
        viewModel.lastName = binding.edtLastName.text.toString ()
        viewModel.age = age
        viewModel.phone = binding.edtPhone.text.toString ()
    }

    override fun onStop() {
        super.onStop()

        val sharedPreferences = getSharedPreferences(registerForm, MODE_PRIVATE)
        sharedPreferences.edit {
            putString ("firstName", viewModel.firstName)
            putString ("lastName", viewModel.lastName)
            putInt ("age", viewModel.age)
            putString ("phone", viewModel.phone)

            apply()
        }
    }

    fun loadPreferences () {
        val sharedPreferences = getSharedPreferences(registerForm, MODE_PRIVATE)
        val firstName = sharedPreferences.getString("firstName", "")
        val lastName = sharedPreferences.getString("lastName", "")
        val age = sharedPreferences.getInt("age", 0)
        val phone = sharedPreferences.getString("phone", "")

        viewModel.firstName = firstName!!
        viewModel.lastName = lastName!!
        viewModel.age = age!!
        viewModel.phone = phone!!
    }

}